src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CodeActionEndpoint.cs

```diff
@@ -290,49 +297,44 @@ internal async Task<IEnumerable<RazorVSInternalCodeAction>> GetCodeActionsFromLa
         return await _languageServer.SendRequestAsync<DelegatedCodeActionParams, RazorVSInternalCodeAction[]>(RazorLanguageServerCustomMessageTargets.RazorProvideCodeActionsEndpoint, delegatedParams, cancellationToken);
     }
 
-    private async Task<IEnumerable<RazorVSInternalCodeAction>?> GetRazorCodeActionsAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+    private async Task<ImmutableArray<RazorVSInternalCodeAction>> GetRazorCodeActionsAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
     {
         cancellationToken.ThrowIfCancellationRequested();
 
-        var tasks = new List<Task<IReadOnlyList<RazorVSInternalCodeAction>?>>();
+        using var _ = ArrayBuilderPool<Task<IReadOnlyList<RazorVSInternalCodeAction>?>>.GetPooledObject(out var tasks);
 
         foreach (var provider in _razorCodeActionProviders)
         {
-            var result = provider.ProvideAsync(context, cancellationToken);
-            if (result != null)
-            {
-                tasks.Add(result);
-            }
+            tasks.Add(provider.ProvideAsync(context, cancellationToken));
         }
 
-        return await ConsolidateCodeActionsFromProvidersAsync(tasks, cancellationToken);
+        return await ConsolidateCodeActionsFromProvidersAsync(tasks.ToImmutableArray(), cancellationToken);
     }
 
-    private static async Task<IEnumerable<RazorVSInternalCodeAction>?> ConsolidateCodeActionsFromProvidersAsync(
-        List<Task<IReadOnlyList<RazorVSInternalCodeAction>?>> tasks,
+    private static async Task<ImmutableArray<RazorVSInternalCodeAction>> ConsolidateCodeActionsFromProvidersAsync(
+        ImmutableArray<Task<IReadOnlyList<RazorVSInternalCodeAction>?>> tasks,
         CancellationToken cancellationToken)
     {
         var results = await Task.WhenAll(tasks).ConfigureAwait(false);
-        var codeActions = new List<RazorVSInternalCodeAction>();
+
+        using var _ = ArrayBuilderPool<RazorVSInternalCodeAction>.GetPooledObject(out var codeActions);
 
         cancellationToken.ThrowIfCancellationRequested();
 
-        for (var i = 0; i < results.Length; i++)
+        foreach (var result in results)
         {
-            var result = results[i];
-
             if (result is not null)
             {
                 codeActions.AddRange(result);
             }
         }
 
-        return codeActions;
+        return codeActions.ToImmutableArray();
     }
 
     private static ImmutableHashSet<string> GetAllAvailableCodeActionNames()
     {
-        var availableCodeActionNames = new HashSet<string>();
+        var _ = ArrayBuilderPool<string>.GetPooledObject(out var availableCodeActionNames);
```

Comment:
Shouldn't this include `using` so that the builder gets reclaimed?

Reply:
Whoops! Will see if there is a VS Threading analyzer we can turn on that catches this